Add Endnote XML Exporter + Rehaul Importer

subhramit opened the pull request:
Closes #11137 
I am attempting to write an Endnote XML Exporter for JabRef.
I have successfully been able to implement the exporter, but there are a few catches, which will hopefully be addressed in follow-up issues and PRs:
1. ~Journal/Booktitle values beginning with "IEEE" are not captured:~
Original Library:
![Screenshot 2024-04-07 145906](https://github.com/JabRef/jabref/assets/74734844/5966adc8-bfe9-4dc3-a1f2-3a17b3f95eb8)
On export and import:
![Screenshot 2024-04-07 150630](https://github.com/JabRef/jabref/assets/74734844/514067c4-33aa-4487-9a01-f8af67c9f247)

2. ~The author/editor of the third entry changes from "Madrigal" to "Marigal and o S." (refer to the same pair of screenshots).~

3. @koppor @Siedlerchr @HoussemNasri @ThiloteE @calixtus ~The code is in a very raw stage right now~, if you have time to go through it and suggest enhancements, please do.

Importer changes:
The importer had to be tweaked a little to include fields such as ```tertiary-title``` corresponding to the Book title/Journal, otherwise the column would remain empty:
![Screenshot 2024-04-07 071355](https://github.com/JabRef/jabref/assets/74734844/108b630d-cd88-4e10-9d73-3924b71afde5)

Similarly, many other fields that were not retrieved by the importer as visible in BibTeX source were added to create a one-to-one mapping between the source entry and the one produced by a round trip export-import. The function ```convertRefNameToType``` was also modified to add some Entrytypes that were not recognized before.

Update: Major changes have taken place, please follow the discussion for details.

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

Siedlerchr commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,218 @@
+        if (authors.isPresent()) {
+            xml.writeStartElement("contributors");
+            xml.writeStartElement("authors");
+            for (String author : authors.get().split("and")) {
```
You should not rely on simply splitting by and. You need to use the AuthorLIstParser, see an example here

https://github.com/JabRef/jabref/blob/a902524861dc7975ec98b62a2dffcd818192380b/src/main/java/org/jabref/logic/msbib/MSBibConverter.java#L145-L147 

Author class itself has methods for getting given (first) or lastname 

The Author class has methods like

subhramit replied:
Done. This solved the author name issue:
![image](https://github.com/JabRef/jabref/assets/74734844/e5e39246-cb21-4fd6-858e-0b91dd9d6f02)

Siedlerchr commented on src/main/java/org/jabref/logic/exporter/ExporterFactory.java:
```diff
@@ -55,6 +55,7 @@ public static ExporterFactory create(PreferencesService preferencesService,
         exporters.add(new TemplateExporter("MIS Quarterly", "misq", "misq", "misq", StandardFileType.RTF, layoutPreferences, saveOrder));
         exporters.add(new TemplateExporter("CSL YAML", "yaml", "yaml", null, StandardFileType.YAML, layoutPreferences, saveOrder, BlankLineBehaviour.DELETE_BLANKS));
         exporters.add(new TemplateExporter("Hayagriva YAML", "hayagrivayaml", "hayagrivayaml", null, StandardFileType.YAML, layoutPreferences, saveOrder, BlankLineBehaviour.DELETE_BLANKS));
+        exporters.add(new TemplateExporter("Endnote XML", "endnotexml", "Endnote XML", "endnotexml", StandardFileType.XML, layoutPreferences, saveOrder));
```
A Template exporter is one that only uses the .layout files -> Remove this line

subhramit replied:
Done.

Siedlerchr commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        try (BufferedWriter writer = Files.newBufferedWriter(file, StandardCharsets.UTF_8)) {
+            XMLOutputFactory xof = XMLOutputFactory.newFactory();
```
Move the factory initialization to the constructor because it's an expensive operation
         XMLOutputFactory outputFactory = XMLOutputFactory.newFactory();

subhramit replied:
Done.

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -207,7 +210,9 @@ private static EntryType convertRefNameToType(String refName) {
             case "book section" -> StandardEntryType.InBook;
             case "book" -> StandardEntryType.Book;
             case "report" -> StandardEntryType.Report;
-            // case "journal article" -> StandardEntryType.Article;
```
is journal article no longer valid?

HoussemNasri commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,209 @@
+            try {
+                xml.writeStartElement("contributors");
+                xml.writeStartElement("authors");
+                for (Author author : authorListParser.parse(authors).getAuthors()) {
```
nitpick: I feel this line is doing a lot of things, can you simplify it a bit?

subhramit replied:
I made a tiny separation, see if this makes it better.

HoussemNasri commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,209 @@
+
+        try (BufferedWriter writer = Files.newBufferedWriter(file, StandardCharsets.UTF_8)) {
+            XMLStreamWriter xml = xmlOutputFactory.createXMLStreamWriter(writer);
+            xml.writeStartDocument(StandardCharsets.UTF_8.name(), "1.0");
```
This would output XML in one line, it's more storage-efficient but it would save you a few KBs or MBs. It's better to output formatted XML for easier debugging.

![image](https://github.com/JabRef/jabref/assets/21198231/63fe3149-12df-4e44-a63d-93e5fe304f0b)

subhramit replied:
Cool, I updated to get formatted XML:
![image](https://github.com/JabRef/jabref/assets/74734844/ed1764bc-f706-476e-9cd5-714ec8e148fa)

Siedlerchr commented:
In general looks good to me. As a follow up task and PR you can think about converting it to a StreamWriter (StaX).
StaX is more memory efficient. See e..g ModsExporter  as an example

subhramit commented:
> In general looks good to me. As a follow up task and PR you can think about converting it to a StreamWriter (StaX). StaX is more memory efficient. See e..g ModsExporter as an example

Yes, in fact I had a couple of plans for the follow up PRs I'm going to put regarding this.
1. Fixing the IEEE booktitle (point 1 of PR description) (might raise this as an issue once this is merged)
2. Writing a test for this
3. StaX as you suggested - will read up on this.

subhramit commented:
@Schaggerlag Fixing the changelog conflict, it's saying I "dismissed" your review. That was accidental.

Siedlerchr commented:
Please do add the test here  well. This makes the migration to stax easier  Because you then have already a working test to verfiy the functionality.

subhramit commented:
> Please do add the test here well. This makes the migration to stax easier Because you then have already a working test to verfiy the functionality.

On it.

subhramit commented:
I am getting

> > Please do add the test here well. This makes the migration to stax easier Because you then have already a working test to verfiy the functionality.
> 
> On it.

I wrote a test (used an LLM for the boilerplate and formatting) - the issue I seem to be facing is that for the following part:
```
<ref-type name="Journal Article">
  <style face="normal" font="default" size="100%">15</style>
</ref-type>
```
I can't simply adapt to the ref-type number (here 15) as the order of entries in the ```EXPORT_ITEM_TYPE``` map is probably changing at every test run, which is causing the test to fail. Any solutions?

subhramit commented:
Okay, I added some more fields to the exporter and updated the importer and test. Now there is a one-to-one mapping between the the fields of the original BibTeX source and the exported one.

subhramit commented:
> > I can't simply adapt to the ref-type number (here 15) as the order of entries in the EXPORT_ITEM_TYPE map is probably changing at every test run, which is causing the test to fail. Any solutions?
> 
> You needd a LinkedHashmap that keeps the insertion order https://www.baeldung.com/java-linked-hashmap

Yes, I manually added the order of insertion (as LinkedHashMap didn't directly allow me to do map.of(...)

Siedlerchr commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,248 @@
+    );
+
+    private static final Map<EntryType, String> EXPORT_REF_NUMBER = EXPORT_ITEM_TYPE.entrySet().stream()
+                                                                                    .collect(Collectors.toMap(
```
https://dkbalachandar.wordpress.com/2017/04/03/java-8-create-linkedhashmap-with-collectors-tomap/

subhramit replied:
Got it, changed.

subhramit replied:
Okay this is causing the test to fail again. I'll look into it in a while.

subhramit replied:
Kkay finally done.

Siedlerchr commented on src/test/java/org/jabref/logic/exporter/EndnoteXmlExporterTest.java:
```diff
@@ -0,0 +1,174 @@
+        Path file = tempDir.resolve("ThisIsARandomlyNamedFile");
+        Files.createFile(file);
+
+        exporter.export(databaseContext, file, Collections.singletonList(entry));
```
This huge xml file here inside is a bit difficult to read. 
1. Create a bib file with the export entry
2. Create the expected xml file
3. Name them with the same name
4. Put them in the folder under `src/test/resources/org/jabref/logic/exporter`
And adapt the test to read them. 
Check out  the MSBibExportFormatFilesTest for example to see how it's done

HoussemNasri commented:
Great job so far! However I noticied a couple of problems with the exported libraries.

After exporting a sizable library and re-importing it into JabRef, I noticed that some authors were missing, despite both instances being sorted by authors in ascending order. The library used can be accessed here: [phys.bib](https://file.io/PvxD8a2iLFXb).

Here are the screenshots for comparison:
- Original: ![original](https://github.com/JabRef/jabref/assets/21198231/3d08d27e-b7d9-46ce-a34f-bc7950fc0519)
- Exported Library: ![exported-library](https://github.com/JabRef/jabref/assets/21198231/f430a49b-b06b-496e-bbd1-60a27ed1f368)

I also attempted to load this [library](https://file.io/UXqtAtHg35kh) which I exported with JabRef into Endnote x8 and 20, and they both crashed. This issue did not occur when exporting with Zotero, indicating a potential problem on our end.

subhramit commented:
> Great job so far! However I noticied a couple of problems with the exported libraries.
> 
> After exporting a sizable library and re-importing it into JabRef, I noticed that some authors were missing, despite both instances being sorted by authors in ascending order. The library used can be accessed here: [phys.bib](https://file.io/PvxD8a2iLFXb).
> 
> Here are the screenshots for comparison:
> 
> * Original: ![original](https://private-user-images.githubusercontent.com/21198231/321014587-3d08d27e-b7d9-46ce-a34f-bc7950fc0519.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI3MDI3MzcsIm5iZiI6MTcxMjcwMjQzNywicGF0aCI6Ii8yMTE5ODIzMS8zMjEwMTQ1ODctM2QwOGQyN2UtYjdkOS00NmNlLWEzNGYtYmM3OTUwZmMwNTE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDIyNDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyYzg5OGUxMTRiOTMyNmQ3ODE0ZWE3NDA0MmU1OWYyOTk0MDUyNTQyNzk3NzY1MTliYzdlNTY4NjYyYjY2MjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.T-DWScOFfWYxc1HL2UOaojeYzWEVdrULeoUbJPAQeuc)
> * Exported Library: ![exported-library](https://private-user-images.githubusercontent.com/21198231/321014621-f430a49b-b06b-496e-bbd1-60a27ed1f368.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI3MDI3MzcsIm5iZiI6MTcxMjcwMjQzNywicGF0aCI6Ii8yMTE5ODIzMS8zMjEwMTQ2MjEtZjQzMGE0OWItYjA2Yi00OTZlLWJiZDEtNjBhMjdlZDFmMzY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDIyNDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwYmVlMWU5ZTI1ZjQyZmFmMTZhOGYzYzY3ZDIzYjYwNjQzZWIzZjkwZWM1NjM3ODI5MWEyNTBkMGNjNmVmNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.96dhEihHOWocoxHRP233_ZDS4K45NwUNJJZ9RyV8me0)
> 
> I also attempted to load this [library](https://file.io/UXqtAtHg35kh) which I exported with JabRef into Endnote x8 and 20, and they both crashed. This issue did not occur when exporting with Zotero, indicating a potential problem on our end.
> 
> ![endnote-import-error](https://private-user-images.githubusercontent.com/21198231/321016742-0b6af2de-1237-417d-a29b-4df8641bfe8d.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI3MDI3MzcsIm5iZiI6MTcxMjcwMjQzNywicGF0aCI6Ii8yMTE5ODIzMS8zMjEwMTY3NDItMGI2YWYyZGUtMTIzNy00MTdkLWEyOWItNGRmODY0MWJmZThkLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDIyNDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzMWUzOTBmYzQ5NTBmZDAxODcwMTMzZjgwZWJlNTY2MmMzN2FlMjkyMGIwYzUyNTFmMzE5NzE3N2Q5OWQyNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2OxU6LFntVGur4EyBMBUFBoTno6dBYtkhLUELUWu2Z8)

Hi Houssem, I tried accessing the phys.bib library but it is showing file deleted. Could you reupload it once?

HoussemNasri commented:
That was weird, it deleted them almost instantly. I uploaded the files on another website. 

This is phys.bib: https://easyupload.io/xvr9i9
This is lib3.xml (the one that crashed Endnote): https://easyupload.io/rz1sqd

subhramit commented:
> That was weird, it deleted them almost instantly. I uploaded the files on another website.
> 
> This is phys.bib: https://easyupload.io/xvr9i9 This is lib3.xml (the one that crashed Endnote): https://easyupload.io/rz1sqd

Thanks! And can you also send the library from which lib3.xml was produced? (Or is it the same phys.bib?)

Siedlerchr commented:
You can upload the bib file here as well in github, you just have to zip it

HoussemNasri commented:
> You can upload the bib file here as well in github, you just have to zip it

Thanks for the tip! One would think I'd be familiar with this after spending so much time on GitHub :-)

subhramit commented:
I'm working on the above two issues. I see why the authors were missing. Will try to push fixes soon.

subhramit commented:
Seems like the importer has to be revamped heavily. It also fails to parse a couple of titles, apart from keywords, editors and so on.
I have an exam coming up on 13th, wo will get back to this on Saturday.

subhramit commented:
> Seems like the importer has to be revamped heavily. It also fails to parse a couple of titles, apart from keywords, editors and so on. I have an exam coming up on 13th, so will get back to this on Saturday.

So far I've been able to fix the authors ("editors", to be precise) issue:
![image](https://github.com/JabRef/jabref/assets/74734844/fa82c541-9e9a-4496-839b-a777b19e565e)

koppor commented:
> I'm working on the above two issues. I see why the authors were missing. Will try to push fixes soon.

Random check: Please also add a test case. Not using the full file, but a minimal example. Maybe even without concrete files and just testing the methods somehow?

subhramit commented:
> > I'm working on the above two issues. I see why the authors were missing. Will try to push fixes soon.
> 
> Random check: Please also add a test case. Not using the full file, but a minimal example. Maybe even without concrete files and just testing the methods somehow?

Sure, I'll add a robust test once I get everything running.

koppor commented:
> > > I'm working on the above two issues. I see why the authors were missing. Will try to push fixes soon.
> > 
> > Random check: Please also add a test case. Not using the full file, but a minimal example. Maybe even without concrete files and just testing the methods somehow?
> 
> Sure, I'll add a robust test once I get everything running.

If you narrow down the issue, you can write the test first and then develop until the test is green. This is called test-driven development and reduces development time.

Remember last section at https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html

subhramit commented:
> > > > I'm working on the above two issues. I see why the authors were missing. Will try to push fixes soon.
> > > 
> > > Random check: Please also add a test case. Not using the full file, but a minimal example. Maybe even without concrete files and just testing the methods somehow?
> > 
> > Sure, I'll add a robust test once I get everything running.
> 
> If you narrow down the issue, you can write the test first and then develop until the test is green. This is called test-driven development and reduces development time.
> 
> Remember last section at https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html

Actually why narrowing down seems a little difficult is because of the second issue that @HoussemNasri pointed out. There may be many adaptations that need to be done to conform to the endnote dtd, I can currently get an idea based on how Zotero's endnote export files look like, but they may not be compatible with our importer presently. Once I can highlight the format, the other issues are easy to narrow down and write a test for in that format...

subhramit commented:
Okay, back to this.
Could someone tell me if the "file" field is necessary while exporting?
![image](https://github.com/JabRef/jabref/assets/74734844/2f781ed3-1e82-41ee-8f78-d9686bbd3162)

Siedlerchr commented:
It depends. How does EndNote store external files?
Or can you just export it as a custom field?

There is a class

FileFieldWriter which does the serialization of the file field


Subhramit Basu Bhowmick ***@***.***> schrieb am Sa., 13. Apr.
2024, 20:12:

> Okay, back to this.
> Could someone tell me if the "file" field is necessary while exporting?
> image.png (view on web)
> <https://github.com/JabRef/jabref/assets/74734844/2f781ed3-1e82-41ee-8f78-d9686bbd3162>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/11157#issuecomment-2053720344>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOFZCJ3HNMD7AW6M7G3UDY5FYQJAVCNFSM6AAAAABF3FZ6WOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJTG4ZDAMZUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

subhramit commented:
> It depends. How does EndNote store external files? Or can you just export it as a custom field? There is a class FileFieldWriter which does the serialization of the file field Subhramit Basu Bhowmick ***@***.***> schrieb am Sa., 13. Apr. 2024, 20:12:
> […](#)
> Okay, back to this. Could someone tell me if the "file" field is necessary while exporting? image.png (view on web) <https://github.com/JabRef/jabref/assets/74734844/2f781ed3-1e82-41ee-8f78-d9686bbd3162> — Reply to this email directly, view it on GitHub <[#11157 (comment)](https://github.com/JabRef/jabref/pull/11157#issuecomment-2053720344)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACOFZCJ3HNMD7AW6M7G3UDY5FYQJAVCNFSM6AAAAABF3FZ6WOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJTG4ZDAMZUGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I just checked the document type definition, there is no explicit field for external files:
```
<!ELEMENT record (database?, source-app?, rec-number?, foreign-keys?, ref-type?, contributors?, auth-address?, auth-affiliaton?, titles?, periodical?, pages?, volume?, number?, issue?, secondary-volume?, secondary-issue?, num-vols?, edition?, section?, reprint-edition?, reprint-status?, keywords?, dates?, pub-location?, publisher?, orig-pub?, isbn?, accession-num?, call-num?, report-id?, coden?, electronic-resource-num?, abstract?, label?, image?, caption?, notes?, research-notes?, work-type?, reviewed-item?, availability?, remote-source?, meeting-place?, work-location?, work-extent?, pack-method?, size?, repro-ratio?, remote-database-name?, remote-database-provider?, language?, urls?, access-date?, modified-date?, custom1?, custom2?, custom3?, custom4?, custom5?, custom6?, custom7?, misc1?, misc2?, misc3?)>
```
However, there are custom fields we might be able to use.

HoussemNasri commented:
External files and URLs are stored in the urls field in Endnote. See how Zotero parses it:
https://github.com/zotero/translators/blob/db2771d52d89d1480ff98efbd6968565893f2184/Endnote%20XML.js#L636-L667

However, you don't have to implement it in this PR if it's too complicated. You can always create a follow-up.

subhramit commented:
Alright, I have been able to fix pretty much every problem that has been discussed above:
1. The exported file format is now along the lines of how it is done by Zotero, and thus compatible with Endnote (tested with Endnote 20)
2. Importer has been adapted accordingly. Round-trip export-import works flawlessly.
3. Missing secondary authors (editors) issue fixed.
4. Missing fields have been added (now it is more or less a one-to-one mapping on inspecting the BibTeX sources).

Only things left are working on tests. The exporter test is failing because of the order of field mappings not being preserved again, will look into it soon. The importer test is probably failing because of the rehaul.
@koppor @Siedlerchr @HoussemNasri I don't think follow-up PRs will be required. Requesting your reviews. 
I will put the xml as a separate file and test using that as suggested by @Siedlerchr once I am done fixing the tests. I primarily wanted to get done with the logic part before my end-semester examinations begin.

Siedlerchr commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,201 @@
+
+public class EndnoteXmlExporter extends Exporter {
+    private static final List<EntryType> ORDERED_ENTRY_TYPES = List.of(
+            StandardEntryType.Article, // Journal Article -1
```
Why not make an enum or a record all the necessary information?
e.g. EndnoteEntryTypes(StandardEntryType jabrefEntryType, String endnoteEntryType, int order)

subhramit replied:
Initially, I was using the LinkedHashMap approach:
```
private static final Map<EntryType, String> ENTRY_TYPE_MAPPING = new LinkedHashMap<>(Map.ofEntries(
            Map.entry(StandardEntryType.Article, "Journal Article"),
            Map.entry(StandardEntryType.Book, "Book"),
            Map.entry(StandardEntryType.InBook, "Book Section"),
            Map.entry(StandardEntryType.InCollection, "Book Section"),
            Map.entry(StandardEntryType.Proceedings, "Conference Proceedings"),
            Map.entry(StandardEntryType.MastersThesis, "Thesis"),
            Map.entry(StandardEntryType.PhdThesis, "Thesis"),
            Map.entry(StandardEntryType.TechReport, "Report"),
            Map.entry(StandardEntryType.Unpublished, "Manuscript"),
            Map.entry(StandardEntryType.InProceedings, "Conference Paper"),
            Map.entry(StandardEntryType.Conference, "Conference"),
            Map.entry(IEEETranEntryType.Patent, "Patent"),
            Map.entry(StandardEntryType.Online, "Web Page"),
            Map.entry(IEEETranEntryType.Electronic, "Electronic Article"),
            Map.entry(StandardEntryType.Misc, "Generic")
    ));

    private static final Map<EntryType, String> EXPORT_REF_NUMBER = ENTRY_TYPE_MAPPING.keySet().stream().collect(
            Collectors.toMap(entryType -> entryType, entryType -> Integer.toString(ENTRY_TYPE_MAPPING.keySet().stream().toList().indexOf(entryType) + 1), (e1, e2) -> e1, LinkedHashMap::new)); 
           
```

But then no matter what I did, the order of reference type number was random. This was just a workaround that finally fixed it. I'll try your suggestion.

subhramit commented:
Update: I am having a hard time with the tests, as my current implementation exports fields in a non-pre-defined order, and thus a test cannot be hardcoded. I'll have to find a way or perhaps fix the order in which fields are written in the xml file, which will require another round of changes in the exporter and importer. Furthermore, from the logs, I am unable to understand why exactly the importer test is failing.

If anybody has any ideas regarding the above, please let me know.

Siedlerchr commented:
I will take a deeper look later. Maybe I have an idea

Siedlerchr commented:
I split the tests now in two separate test and moved it to the resources directory
I also added a roundtrip test bib entry- >  exporter -> importer -> bib entry

subhramit commented:
> I split the tests now in two separate test and moved it to the resources directory I also added a roundtrip test bib entry- > exporter -> importer -> bib entry
> 
> Used the xml matcher which ignores the field order

Thank you. This will help immensely.
On my part, I had done some modifications to preserve field order (my last commit above this comment). Right now, on inspecting the dtd and zotero's format a little more carefully, there were a few more disparities that neede to be resolved (like ```year``` being a child of ```dates```, etc.
The importer will need changes, as it is no longer able to detect it as a valid format.
I will adapt the tests accordingly. I will be pushing my last commit after this comment, let me know if something could be simplified.

Siedlerchr commented:
If you want to re-use the entry types and the field mapping for the  importer and exporter you can take a look at Guava's BiMap , see e.g. MSBibMapping 
https://github.com/JabRef/jabref/blob/a902524861dc7975ec98b62a2dffcd818192380b/src/main/java/org/jabref/logic/msbib/MSBibMapping.java#L187-L194

subhramit commented:
Work on importer and exporter is done. They work flawlessly within JabRef and across Zotero/Endnote now.
Original:
![Screenshot 2024-04-15 080925](https://github.com/JabRef/jabref/assets/74734844/35b2ccf4-edba-4e0e-a996-9f18362f50be)
On round-trip export-import:
![Screenshot 2024-04-15 080938](https://github.com/JabRef/jabref/assets/74734844/5c15ee32-d89f-4eb2-8ca7-e8a332c6b239)

There is however, just one problem which, even after a lot of investigation, I couldn't fix - "journaltitle" field maps to "journal" on import.
Original:
![Screenshot 2024-04-15 080837](https://github.com/JabRef/jabref/assets/74734844/692630fb-9c4a-4e5a-adba-38b024def037)
On round-trip export-import:
![Screenshot 2024-04-15 080849](https://github.com/JabRef/jabref/assets/74734844/bd12ef96-461c-461a-b0e2-6a716cf23a71)

This might create an issue with regards to round-trip tests.
Pending: 1. Adapting individual tests (put in the resources folder by @Siedlerchr ) 2. Enhance re-use of field mappings of importer and exporter, based on MSBibMapping's implementation.

Siedlerchr commented:
Regarding journal and journaltitle, the first is bibtex, the second one is
biblatex
There are methods like getFieldOrAlias for such fields, useful for
exporting

and I think there is somewhere an import cleanup operation which takes care
of this as well after import

Subhramit Basu Bhowmick ***@***.***> schrieb am Mo., 15. Apr.
2024, 04:56:

> Work on importer and exporter is complete. They work flawlessly within
> JabRef and across Zotero/Endnote now.
> Original:
> Screenshot.2024-04-15.080925.png (view on web)
> <https://github.com/JabRef/jabref/assets/74734844/35b2ccf4-edba-4e0e-a996-9f18362f50be>
> On round-trip export-import:
> Screenshot.2024-04-15.080938.png (view on web)
> <https://github.com/JabRef/jabref/assets/74734844/5c15ee32-d89f-4eb2-8ca7-e8a332c6b239>
>
> There is however, just one problem which, even after a lot of
> investigation, I couldn't fix - "journaltitle" field maps to "journal" on
> import.
> Original:
> Screenshot.2024-04-15.080837.png (view on web)
> <https://github.com/JabRef/jabref/assets/74734844/692630fb-9c4a-4e5a-adba-38b024def037>
> On round-trip export-import:
> Screenshot.2024-04-15.080849.png (view on web)
> <https://github.com/JabRef/jabref/assets/74734844/bd12ef96-461c-461a-b0e2-6a716cf23a71>
>
> This might create an issue with regards to round-trip tests.
> Pending: 1. Adapting individual tests (put in the resources folder by
> @Siedlerchr <https://github.com/Siedlerchr> ) 2. Enhance re-use of field
> mappings of importer and exporter, based on MSBibMapping's implementation.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/11157#issuecomment-2054563990>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOFZGYOV4VFTOCFCTXIG3Y5M6WFAVCNFSM6AAAAABF3FZ6WOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJUGU3DGOJZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

subhramit commented:
> Regarding journal and journaltitle, the first is bibtex, the second one is biblatex There are methods like getFieldOrAlias for such fields, useful for exporting and I think there is somewhere an import cleanup operation which takes care of this as well after import Subhramit Basu Bhowmick ***@***.***> schrieb am Mo., 15. Apr. 2024, 04:56:
> […](#)
> Work on importer and exporter is complete. They work flawlessly within JabRef and across Zotero/Endnote now. Original: Screenshot.2024-04-15.080925.png (view on web) <https://github.com/JabRef/jabref/assets/74734844/35b2ccf4-edba-4e0e-a996-9f18362f50be> On round-trip export-import: Screenshot.2024-04-15.080938.png (view on web) <https://github.com/JabRef/jabref/assets/74734844/5c15ee32-d89f-4eb2-8ca7-e8a332c6b239> There is however, just one problem which, even after a lot of investigation, I couldn't fix - "journaltitle" field maps to "journal" on import. Original: Screenshot.2024-04-15.080837.png (view on web) <https://github.com/JabRef/jabref/assets/74734844/692630fb-9c4a-4e5a-adba-38b024def037> On round-trip export-import: Screenshot.2024-04-15.080849.png (view on web) <https://github.com/JabRef/jabref/assets/74734844/bd12ef96-461c-461a-b0e2-6a716cf23a71> This might create an issue with regards to round-trip tests. Pending: 1. Adapting individual tests (put in the resources folder by @Siedlerchr <https://github.com/Siedlerchr> ) 2. Enhance re-use of field mappings of importer and exporter, based on MSBibMapping's implementation. — Reply to this email directly, view it on GitHub <[#11157 (comment)](https://github.com/JabRef/jabref/pull/11157#issuecomment-2054563990)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACOFZGYOV4VFTOCFCTXIG3Y5M6WFAVCNFSM6AAAAABF3FZ6WOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJUGU3DGOJZGA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

The export field is as expected, and I tried the cleanup operation "convert to biblatex" after importing - it indeed becomes journaltitle. Thanks.

koppor commented:
> I tried the cleanup operation "convert to biblatex" after importing -

I did not follow the full conversation.

Reading this, it is about "biblatex"  mode. However, if a user has a "bibtex" library, the import should use `journal` (thus, nonono convert to biblatex cleanup)

@Siedlerchr I think, we discussed somewhere about a "nice" fetcher infrasturcture. Each fetcher should create bibtex and we convert to biblatex after the fetcher. I found https://github.com/JabRef/jabref/pull/1594 only, but not the discussion. I think, the argument (of this unfound discussion) could also apply here?

subhramit commented:
What is the appropriate StandardField for ```accession-number```?

subhramit commented:
> > I tried the cleanup operation "convert to biblatex" after importing -
> 
> I did not follow the full conversation.
> 
> Reading this, it is about "biblatex" mode. However, if a user has a "bibtex" library, the import should use `journal` (thus, nonono convert to biblatex cleanup)
> 
> @Siedlerchr I think, we discussed somewhere about a "nice" fetcher infrasturcture. Each fetcher should create bibtex and we convert to biblatex after the fetcher. I found #1594 only, but not the discussion. I think, the argument (of this unfound discussion) could also apply here?

So do I need to do anything about this?

subhramit commented:
I wish to add two more fields which I found in a test file:
![image](https://github.com/JabRef/jabref/assets/74734844/785e8055-29ae-409a-9ffd-c524cdae7b67)

What should be the appropriate StandardField for user-16 and accession-number? I have already used StandardField.NUMBER for "number" so the map won't allow duplicate entries.

Siedlerchr commented:
> @Siedlerchr I think, we discussed somewhere about a "nice" fetcher infrasturcture. Each fetcher should create bibtex and we convert to biblatex after the fetcher. I found #1594 only, but not the discussion. I think, the argument (of this unfound discussion) could also apply here?

https://github.com/JabRef/jabref/blob/0d1b5e8a6310591db8e298eebf3c9181f54090f5/src/main/java/org/jabref/logic/importer/ImportCleanup.java#L12-L13

This import cleanup is always run after importing (fetcher or normal import (Import Handler). That is already covered.

https://github.com/JabRef/jabref/blob/0d1b5e8a6310591db8e298eebf3c9181f54090f5/src/main/java/org/jabref/gui/JabRefFrame.java#L551-L553

subhramit commented:
> I wish to add two more fields which I found in a test file: ![image](https://private-user-images.githubusercontent.com/74734844/322494804-785e8055-29ae-409a-9ffd-c524cdae7b67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxOTAxNzMsIm5iZiI6MTcxMzE4OTg3MywicGF0aCI6Ii83NDczNDg0NC8zMjI0OTQ4MDQtNzg1ZTgwNTUtMjlhZS00MDlhLTlmZmQtYzUyNGNkYWU3YjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE1VDE0MDQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2MzE5NDE1MzM4ODRkNjRiNzhiZjQ2ZjhhZjU1NzRhYmZmZDNjZTlmM2JiY2E5ZmYxOTRhM2JiMTMwMjU3OGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.37ekmaXsg3R13Rf0bs-Y8cLaDuhImQt9dektGkm-FA0)
> 
> What should be the appropriate StandardField for "user-16" and "accession-num"?

Okay never mind, these are not bibtex fields, and the old importer could import such xml entries by skipping them.

Siedlerchr commented:
> > I wish to add two more fields which I found in a test file: ![image](https://private-user-images.githubusercontent.com/74734844/322494804-785e8055-29ae-409a-9ffd-c524cdae7b67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxOTAxNzMsIm5iZiI6MTcxMzE4OTg3MywicGF0aCI6Ii83NDczNDg0NC8zMjI0OTQ4MDQtNzg1ZTgwNTUtMjlhZS00MDlhLTlmZmQtYzUyNGNkYWU3YjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE1VDE0MDQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2MzE5NDE1MzM4ODRkNjRiNzhiZjQ2ZjhhZjU1NzRhYmZmZDNjZTlmM2JiY2E5ZmYxOTRhM2JiMTMwMjU3OGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.37ekmaXsg3R13Rf0bs-Y8cLaDuhImQt9dektGkm-FA0)
> > What should be the appropriate StandardField for "user-16" and "accession-num"?
> 
> Okay never mind, these are not bibtex fields, and the old importer could import xml entries skipping them.

You can import such fields as `new UnknownField("user-16")`

subhramit commented:
Update: Fixed Importer for failing tests. They were failing because the new importer I wrote wasn't handling the cases when the text field content is embedded within a child ```style``` tag.

subhramit commented:
There is a failing "architecture" test in the logs, though:
```
org.jabref.architecture.TestArchitectureTest

  Test testNaming FAILED

  java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'classes that are top level classes and do not have fully qualified name 'org.jabref.benchmarks.Benchmarks' and do not have fully qualified name 'org.jabref.http.server.TestBibFile' and do not have fully qualified name 'org.jabref.gui.autocompleter.AutoCompleterUtil' and do not have fully qualified name 'org.jabref.gui.search.TextFlowEqualityHelper' and do not have fully qualified name 'org.jabref.logic.bibtex.BibEntryAssert' and do not have fully qualified name 'org.jabref.logic.importer.fileformat.ImporterTestEngine' and do not have fully qualified name 'org.jabref.logic.l10n.JavaLocalizationEntryParser' and do not have fully qualified name 'org.jabref.logic.l10n.LocalizationEntry' and do not have fully qualified name 'org.jabref.logic.l10n.LocalizationParser' and do not have fully qualified name 'org.jabref.logic.openoffice.style.OOBibStyleTestHelper' and do not have fully qualified name 'org.jabref.logic.shared.TestManager' and do not have fully qualified name 'org.jabref.model.search.rules.MockSearchMatcher' and do not have fully qualified name 'org.jabref.model.TreeNodeTestData' and do not have fully qualified name 'org.jabref.performance.BibtexEntryGenerator' and do not have fully qualified name 'org.jabref.support.DisabledOnCIServer' and do not have fully qualified name 'org.jabref.support.CIServerCondition' and do not have fully qualified name 'org.jabref.testutils.interactive.styletester.StyleTesterMain' and do not have fully qualified name 'org.jabref.testutils.interactive.styletester.StyleTesterView' should have simple name ending with 'Test'' was violated (1 times):
  Class <org.jabref.logic.exporter.EndnoteXmlExporterTestFiles> does not have simple name ending with 'Test' in (EndnoteXmlExporterTestFiles.java:0)
  ```

subhramit commented:
Exporter test is failing because it expects:
```
<[@conference{Dai2018,
      abstract = {An episode........
```
While it was
```
<[@conference{,
      abstract = {An episode.......
```
How to fix this?

koppor commented:
> There is a failing "architecture" test in the logs, though:
> [...]
> How to fix this?

Take the time to think about. Much to learn!

Here some guidance.

- Test method ends with `.should().haveSimpleNameEndingWith("Test")`
- Your class is named `EndnoteXmlExporterTestFiles` does NOT end with `Test`
- Let's check other classes. I know there are importers. I checked the sub packages. I found `BiblioscapeImporterFilesTest`.
- I realize, that it ends with `FilesTest`.
- Your class ends with `TestFiles`.

Does that give enough clues?

koppor commented:
> Exporter test is failing because it expects:

There is domain knowledge involved.

Did you read a bit about BibTeX?

Can you tell me the difference between the two entries?

> 
> ```
> <[@conference{Dai2018,
>       abstract = {An episode........
> ```
> 
> While it was
> 
> ```
> <[@conference{,
>       abstract = {An episode.......
> ```
> 
> How to fix this?

...

...

...

OK, I tell you. One has a BibTeX key, the other not.

I do not know the endnote format. Does it have the concept of citation keys? Or the concept of custom fields so that one could store the citation key? If not, then you will need to remove the citation key from the test .bib files probably?

subhramit commented:
Importer tests are failing because the new importer is reading more fields than are expected in the test. This means there are issues with the bib files being used for testing. Does this mean I have to export the xml's and change the bib files?
Apart from that, the order is also different from expected (which again, should not matter).
I might need some more help as these tests are painful :(

subhramit commented:
> > Exporter test is failing because it expects:
> 
> There is domain knowledge involved.
> 
> Did you read a bit about BibTeX?
> 
> Can you tell me the difference between the two entries?
> 
> > ```
> > <[@conference{Dai2018,
> >       abstract = {An episode........
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > While it was
> > ```
> > <[@conference{,
> >       abstract = {An episode.......
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How to fix this?
> 
> ...
> 
> ...
> 
> ...
> 
> OK, I tell you. One has a BibTeX key, the other not.
> 
> I do not know the endnote format. Does it have the concept of citation keys? Or the concept of custom fields so that one could store the citation key? If not, then you will need to remove the citation key from the test .bib files probably?

Yes, I read up while I was making my proposal, but did not know the significance of BibTeX keys. Thank you for telling me. I'll take a look.

> > There is a failing "architecture" test in the logs, though:
> > [...]
> > How to fix this?
> 
> Take the time to think about. Much to learn!
> 
> Here some guidance.
> 
> * Test method ends with `.should().haveSimpleNameEndingWith("Test")`
> * Your class is named `EndnoteXmlExporterTestFiles` does NOT end with `Test`
> * Let's check other classes. I know there are importers. I checked the sub packages. I found `BiblioscapeImporterFilesTest`.
> * I realize, that it ends with `FilesTest`.
> * Your class ends with `TestFiles`.
> 
> Does that give enough clues?

That was my first thought, but since it was added by @Siedlerchr I thought the issue must be something deeper. Thanks, will change it.

Siedlerchr commented:
Simply didn't know about the architecture rule.
Regarding the bibtex format some good description can be found here  https://www.bibtex.org/

subhramit commented:
Importer tests are failing because the new importer is reading more fields than are expected in the test. This means there are issues with the bib files being used for testing. Does this mean I have to export the xml's and change the bib files?
Apart from that, the order is also different from expected (which again, should not matter).
And, and, the closing bracket for Notes is with a newline.
Furthermore, in one case the "number" being imported is the "accession-num", whereas the expected one was the one from the "number" field. My code seems to be fine with regards to these, any clue?

koppor commented:
> Importer tests are failing because the new importer is reading more fields (which should be) than are expected in the test. This means there are issues with the bib files being used for testing. Does this mean I have to export the xml's and change the bib files? Apart from that, the order is also different from expected (which again, should not matter). And, and, the closing bracket for Notes is with a newline. Furthermore, in one case the "number" being imported is the "accession-num", whereas the expected one was the one from the "number" field. My code seems to be fine with regards to these, any clue?

I have no real clue without thinking more.

My thoughts

- are athe bib files crafted for endnote only or are they used double?
- if double, duplicate the fiels, so that one bib for one test 
- i don't know about the bib files itself. are they user-supplied? Then, they should not be changed.
- why is the bibtex key missing now? Was it available before your PR? Then, you did something wrong!
- do not let end field content with newlines. use `trim()`
- i don't know `accession-num`. What is this? Why cannot `number` be used?

subhramit commented:
> > Importer tests are failing because the new importer is reading more fields (which should be) than are expected in the test. This means there are issues with the bib files being used for testing. Does this mean I have to export the xml's and change the bib files? Apart from that, the order is also different from expected (which again, should not matter). And, and, the closing bracket for Notes is with a newline. Furthermore, in one case the "number" being imported is the "accession-num", whereas the expected one was the one from the "number" field. My code seems to be fine with regards to these, any clue?
> 
> I have no real clue without thinking more.
> 
> My thoughts
> 
> * are athe bib files crafted for endnote only or are they used double?

The Importer test bib files were not created by me, nor touched by me during this PR. I have no clue what they were crafted for...
> * if double, duplicate the fiels, so that one bib for one test
> * i don't know about the bib files itself. are they user-supplied? Then, they should not be changed.

Again, no clue, they were already there.
> * why is the bibtex key missing now? Was it available before your PR? Then, you did something wrong!

The issue with the BibteX keys is confined to the exporter, hence the failing tests of the exporter are on me. For the importer, though, as I said, I never touched the test files. Even the expected output of the importer test has missing citation keys!
> * do not let end field content with newlines. use `trim()`

Okay, will do it. Although I'm not sure why this is only happening with notes.
> * i don't know ```accession-num```. What is this? Why cannot `number` be used?

I am importing both `number` and `accession-number` from the endnotes, as it was part of the dtd, but I have no idea why ```accession-num``` is getting priority and being imported as ```number``` instead.

Siedlerchr commented:
As you worked on the importer to import more stuff, it's clear that the existing bib files are now not enough. 
I see the following issue in the failing tests for the importer:

Title is not imported correctly:
See the diff (when you execute the test in intellij, you can see a diff for the failure)
<img width="1509" alt="grafik" src="https://github.com/JabRef/jabref/assets/320228/25ab8cc2-2602-4587-bb53-2988930f9fd9">

subhramit commented:
> As you worked on the importer to import more stuff, it's clear that the existing bib files are now not enough. I see the following issue in the failing tests for the importer:
> 
> Title is not imported correctly: See the diff (when you execute the test in intellij, you can see a diff for the failure) <img alt="grafik" width="1509" src="https://private-user-images.githubusercontent.com/320228/322968493-25ab8cc2-2602-4587-bb53-2988930f9fd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyOTQ3NzEsIm5iZiI6MTcxMzI5NDQ3MSwicGF0aCI6Ii8zMjAyMjgvMzIyOTY4NDkzLTI1YWI4Y2MyLTI2MDItNDU4Ny1iYjUzLTI5ODg5MzBmOWZkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNlQxOTA3NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMjI1OTIyZDU4OTg5ZDZkODE2NzkxMGEyYzk2ZmFjMTExN2MxMTc4M2M4ZjA2NTgyOGZhZTJjMTc3MjE1YTVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Qe5Ryg8kerLfkQ75UDfh3LQOrvpJ1CYAg_0OrT9C5w8">

Seems like ```secondary-title``` is taking precedence over ```title``` again and getting presented as ```title```, just like the confusion between ```accession-num``` and ```number```. The only thing I have in mind is to remove these fields, as I don't know where the presentation control is happening...

subhramit commented:
> I do not know the endnote format. Does it have the concept of citation keys? Or the concept of custom fields so that one could store the citation key? If not, then you will need to remove the citation key from the test .bib files probably?

I checked the Endnote format, it does not have citation keys. I checked the Zotero exporter, it doesn't have it as well, nor did the original importer tests have it. So I am removing them from the .bib files for now. If this is much needed, we can put up a follow-up PR.

subhramit commented:
On checking the importer test files, three or four of the seven had an additional newline character in ```notes``` field. So our importer was working as expected in this regard. Updated the cases.

subhramit commented:
Summary of my last few commits:
1. New Importer is now robust in handling newlines, extra spaces, multi-part titles, etc.
2. Fixed overwriting of "Number" and "Title" fields by "accession-num" and "alt-title".
3. Changed importer tests as per the diff.
More info: After points 1 and 2, the diff was:

Test [1] EndnoteXmlImporterTestArticle2.xml
- Actual entries have additional fields: "accession-num", "address", "alt-title", "date", "journaltitle", "key", "language"

Test [2] EndnoteXmlImporterTest_WithoutUrlStyle2.xml
- Actual entries have additional fields: "accession-num", "address", "alt-title", "date", "journaltitle", "key", "language"

Test [3] EndnoteXmlImporterTest_EmptyKeywordStyle.xml
- Expected had "endnote-label" field, actual has "label" field
- Actual has additional fields: "accession-num", "address", "date", "key"
Test [4] EndnoteXmlImporterTest_WithoutUrlStyle.xml
- Actual entry has additional fields: "accession-num", "address", "date"

Test [5] EndnoteXmlImporterTestReport.xml

- Expected entry type is "report", but actual entry type is "techreport"
- Expected has "endnote-label" field, actual has "label" field
- Actual has additional "key" field

Test [6] EndnoteXmlImporterTestLabelAndMultiTitle.xml
- Actual entries have additional fields: "accession-num", "address", "alt-title", "date", "journaltitle", "key", "language"
- Third actual entry has "journaltitle" field, expected does not
- Third expected entry has "endnote-label" field, actual has "label" field

Test [7] EndnoteXmlImporterTestArticle.xml
- Actual entry has additional fields: "accession-num", "address", "date"

To filter out what fields needed to be added to the bib files, I asked Claude which of the above were Standard BibTeX/BibLateX fields. Turns out only "address", "date", "month", "year", "day", "journaltitle" and "language" were. 
Hence:

4. Removed non-standard fields from the Importer, and adapted the importer bib tests by adding the necessary fields and removing the non-standard (e.g. - endnote-label). The adaption also took care of the order of imports, which made the tests pass, however, this should not matter. @Siedlerchr I'd request you to change the importer test as well to ignore the order of entries (or could be a follow-up PR?)
5. Removed trailing ":PDF" and leading ":" from the file path of the tests.

All the importer, exporter and round-trip tests are green now.

This was my largest PR so far, and although the primary purpose was just to close the linked issue by making an exporter, there were many turns and I learnt a lot. I will be taking some time off to prepare for my university tests that are coming up.
@Siedlerchr @koppor @HoussemNasri Please let me know if there is any feedback. If there are any other major changes needed, I will be doing them after a short break (or possibly in follow-up PRs). For now, I am moving this PR from draft to ready.

subhramit commented:
Oh also, 
6. Fixed a typo in a changelog entry (not associated with this PR) :)

koppor reviewed (commented):
Thank you for providing tests. That made it easy to improve code quality

- One uses methods to group things. Then, some comments become obsolete.
- With some JabRef experience, one gets to know existing helper method (e.g., for keywords and author list parsing). I made use of them in the code.

I saw some things in the EndnoteXmlImporter, which I should handle, too.

koppor commented:
> * Actual entries have additional fields: "accession-num", "address", "alt-title", "date", "journaltitle", "key", "language"

- `alt-title` was not handled. I kept it as TODO
-  `journaltitle` is an alias for `journal` (and only available in BibLaTeX). `journal` works for both BibTeX and BibLaTeX. Therefore, I removed `journaltitle`

koppor commented:
> Oh also, 6. Fixed a typo in a changelog entry (not associated with this PR) :)

File a separate PR. Usually, PRs take longer as thought. Then, separate PRs are more useful.

https://gitbutler.com/ could be helpful.

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -28,42 +26,75 @@
-        // not suported by aalto-xml
-        // xmlInputFactory.setProperty(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
-        // required for reading Unicode characters such as &#xf6;
-        xmlInputFactory.setProperty(XMLInputFactory.IS_COALESCING, true);
```
Did you remove that statement on purpose?

subhramit replied:
Oh, I missed this. Basically, I was having a hard time extending the old importer so I started writing from scratch. In that, I did not think about needing this.

koppor replied:
Then, please re-add this. Also the lines above including "not supported by aalto-xml". Someone in future should have a look at that one.

subhramit replied:
Okay, done.

koppor commented on src/test/resources/org/jabref/logic/importer/fileformat/EndnoteXmlImporterTestArticle2.xml:
```diff
@@ -77,8 +77,7 @@
             <notes>
-                <style face="normal" font="default" size="100%">some notes
-                </style>
+                <style face="normal" font="default" size="100%">some notes</style>
```
Not sure if we should modify the imported files. - Since you do a roundtrip test, is is OKish...

subhramit replied:
Yes, I made sure the old case is also handled, though. We can go back to it if you wish. Initially I had thought it's by mistake.

koppor commented:
@subhramit TLDR for my code changes:

- The use of StringBuilder was strange. I fixed it.
- StringJoiner is the thing you were looking for joining authors (using ` and `)

subhramit commented:
> * `alt-title` was not handled. I kept it as TODO

The exporter, you mean? Right, must've missed it.
In the importer, though, I purposely commented it out as it was replacing "title" on preview.

subhramit commented on src/main/java/org/jabref/logic/exporter/EndnoteXmlExporter.java:
```diff
@@ -0,0 +1,288 @@
+
+            // Map entry type
+            EntryType entryType = entry.getType();
+            String refType = ENTRY_TYPE_MAPPING_LIST.stream()
```
Thanks tons for taking over!
Will remember that. I think this intuition people naturally build with experience, I want to get there...

koppor commented:
> In the importer, though, I purposely commented it out as it was replacing "title" on preview.

It should not, it shold put it into `alt-title` field.

subhramit commented:
> As you worked on the importer to import more stuff, it's clear that the existing bib files are now not enough. I see the following issue in the failing tests for the importer:
> 
> Title is not imported correctly: See the diff (when you execute the test in intellij, you can see a diff for the failure) <img alt="grafik" width="1509" src="https://private-user-images.githubusercontent.com/320228/322968493-25ab8cc2-2602-4587-bb53-2988930f9fd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNTM4MDksIm5iZiI6MTcxMzM1MzUwOSwicGF0aCI6Ii8zMjAyMjgvMzIyOTY4NDkzLTI1YWI4Y2MyLTI2MDItNDU4Ny1iYjUzLTI5ODg5MzBmOWZkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxN1QxMTMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDBmMGJlNTJiMmE4ZDVjNjUyNTY4NzYwMjM5ZmRiOGUxMTBiNzE3ZGM4ZTY0OGM0OTkzNmViNTVjYjNkYzFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9mU2HozGT2LaiSCCY-DT-_UDh1E3Soh6SmBpumJWeO8">

@koppor This is the issue that was occurring with ```alt-title``` and ```accession-num```. I uncommented import of ```alt-title``` to re-check.

subhramit commented:
> > Oh also, 6. Fixed a typo in a changelog entry (not associated with this PR) :)
> 
> File a separate PR. Usually, PRs take longer as thought. Then, separate PRs are more useful.
> 
> https://gitbutler.com/ could be helpful.

Seems like the typo has been fixed in upstream main. Skipping.

subhramit commented:
> > As you worked on the importer to import more stuff, it's clear that the existing bib files are now not enough. I see the following issue in the failing tests for the importer:
> > Title is not imported correctly: See the diff (when you execute the test in intellij, you can see a diff for the failure) <img alt="grafik" width="1509" src="https://private-user-images.githubusercontent.com/320228/322968493-25ab8cc2-2602-4587-bb53-2988930f9fd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNTM4MDksIm5iZiI6MTcxMzM1MzUwOSwicGF0aCI6Ii8zMjAyMjgvMzIyOTY4NDkzLTI1YWI4Y2MyLTI2MDItNDU4Ny1iYjUzLTI5ODg5MzBmOWZkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxN1QxMTMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDBmMGJlNTJiMmE4ZDVjNjUyNTY4NzYwMjM5ZmRiOGUxMTBiNzE3ZGM4ZTY0OGM0OTkzNmViNTVjYjNkYzFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9mU2HozGT2LaiSCCY-DT-_UDh1E3Soh6SmBpumJWeO8">
> 
> @koppor This is the issue that was occurring with `alt-title` and `accession-num`. I uncommented import of `alt-title` to re-check.

Okay you're right. ```alt-title``` is being imported separately. I'll adapt the expected values.

subhramit commented:
Export & Import for ```alt-title``` added and tests adapted accordingly.

koppor commented on src/test/resources/org/jabref/logic/importer/fileformat/EndnoteXmlImporterTestLabelAndMultiTitle.bib:
```diff
@@ -1,5 +1,6 @@
 @article{,
       address = {Cent States Forest Expt Stn, Columbus, OH USA},
+      alt-title = {Ecology},
```
improvement: if `alt-title` matches `journal`, then remove `alt-title`. Maybe, a cleanup at the end.

subhramit replied:
Should the same apply if ```booktitle``` matches ```alt-title```?

subhramit replied:
Yes, I have done that.

koppor commented on src/test/resources/org/jabref/logic/importer/fileformat/EndnoteXmlImporterTestLabelAndMultiTitle.bib:
```diff
@@ -15,6 +16,7 @@ @article{
 }
 @book{,
       address = {Univ Chicago, Chicago, IL USA},
+      alt-title = {Ecology},
```
(comment for below)

Here, I see `journal` set. This is wrong. If the entry is not an `article`, that field should go into `booktitle` or another field. Please try to check the table https://en.wikibooks.org/wiki/LaTeX/Bibliography_Management#BibTeX and think on a case-by-case basis. As fallback, just use `booktitle` before loosing the field.

subhramit replied:
In the original import test xml, it was present as ```alt-title```.
Somewhere I had read during all this that ```booktitle``` in the bib is handled by ```secondary-title``` in the Endnote xml format. So I'll modify the bib and the xml, and change the StandardField of secondary-type to BOOKTITLE.

subhramit replied:
Nope, have to undo this. It's importing as booktitle even for articles. Will try to do it the way you suggested.

koppor commented:
Review hint: The diff between two commits can be retrieved by URL hacking: https://github.com/JabRef/jabref/compare/60f3ae2...0527ec7

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -125,366 +160,300 @@ public ParserResult importDatabase(BufferedReader input) throws IOException {
                     }
-                    case "urls" -> {
-                        handleUrlList(reader, fields, linkedFiles);
+                    case "periodical", "alt-periodical" -> {
```
With this, the order `periodical` and `alt-periodical` matters. Just leave out `alt-periodical` - or store it in the field `alt-periodical` (and keep it if it differs from `periodical`).

subhramit replied:
Ah, found it! It was getting imported in the default case. Fixed it.

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -125,366 +160,300 @@ public ParserResult importDatabase(BufferedReader input) throws IOException {
-    }
+        // Cleanup: Remove alt-title if it matches the journal
+        String journalOrBooktitle = entry.getField(StandardField.JOURNAL).or(() -> entry.getField(StandardField.BOOKTITLE)).orElse("");
+        if (entry.hasField(new UnknownField("alt-title"))) {
```
exctract `new UnknownField("alt-title")` to class constant `FIELD_ALT_TITLE`. You can use intellij refactoring extract field - <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd>.

subhramit replied:
Done.

koppor reviewed (changes requested):
Minor comments
